File: algorithms/java/90_subsets_2/Subsets.java
class Subsets {
	// 1. add or not add
	public List<List<Integer>> subsets1(int[] nums) {
		List<List<Integer>> res = new ArrayList<>();
		if (nums == null || nums.length == 0) {
			return res;
		}
		Arrays.sort(nums);
		List<Integer> partial = new ArrayList<>();
		dfs1(nums, res, partial, 0);
		return res;
	}
	private void dfs1(int[] nums, List<List<Integer>> res, List<Integer> partial, int index) {
		if (index == nums.length) {
			res.add(new ArrayList<>(partial));
			return;
		}
		partial.add(nums[index]);
		dfs1(nums, res, partial, index + 1);
		partial.remove(partial.size() - 1);
		while (index + 1 < nums.length && nums[index] == nums[index + 1]) {
			index++;
		}
		dfs1(nums, res, partial, index + 1);
	}

	// 2. combination
	public List<List<Integer>> subsets2(int[] nums) {
		List<List<Integer>> res = new ArrayList<>();
		if (nums == null || nums.length == 0) {
			return res;
		}
		Arrays.sort(nums);
		List<Integer> partial = new ArrayList<>();
		dfs2(nums, res, partial, 0);
		return res;
	}
	private void dfs2(int[] nums, List<List<Integer>> res, List<Integer> partial, int index) {
		res.add(new ArrayList<>(partial));
		for (int i = index; i < nums.length; i++) {
			if (i == index || nums[i] != nums[i - 1]) {
				partial.add(nums[i]);
				dfs2(nums, res, partial, i + 1);
				partial.remove(partial.size() - 1);
			}
		}
	}
}

File: algorithms/cpp/78_subsets/Subsets.cpp
class Subsets {
public:
	vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> cur;
        dfs(res, cur, nums, 0);
        return res;
    }
private:
	void dfs(vector<vector<int>>& res, vector<int>& cur, vector<int>& nums, int index) {
        if (index == nums.size()) {
            res.push_back(cur);
            return;
        }
        dfs(res, cur, nums, index + 1);
        cur.push_back(nums[index]);
        dfs(res, cur, nums, index + 1);
        cur.pop_back();
    }
};
